pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/MessageDeduplication.java

```diff
@@ -0,0 +1,383 @@
+    // after the messages are persisted
+    private final ConcurrentOpenHashMap<String, Long> highestSequencedPersisted = new ConcurrentOpenHashMap<>(16, 1);
+
+    // Number of persisted entries after which to store a snapshot of the sequence ids map
+    private final int snapshotInterval;
+
+    // Counter of number of entries stored after last snapshot was taken
+    private int snapshotCounter;
+
+    // Max number of producer for which to persist the sequence id information
+    private final int maxNumberOfProducers;
+
+    // Map used to track the inactive producer along with the timestamp of their last activity
+    private final Map<String, Long> inactiveProducers = new HashMap<>();
+
+    public MessageDeduplication(PulsarService pulsar, PersistentTopic topic, ManagedLedger managedLedger) {
+        this.pulsar = pulsar;
+        this.topic = topic;
+        this.managedLedger = managedLedger;
+        this.status = Status.Disabled;
+        this.snapshotInterval = pulsar.getConfiguration().getBrokerDeduplicationEntriesInterval();
+        this.maxNumberOfProducers = pulsar.getConfiguration().getBrokerDeduplicationMaxNumberOfProducers();
+        this.snapshotCounter = 0;
+    }
+
+    private CompletableFuture<Void> recoverSequenceIdsMap() {
+        // Load the sequence ids from the snapshot in the cursor properties
+        managedCursor.getProperties().forEach((k, v) -> {
+            highestSequencedPushed.put(k, v);
+            highestSequencedPersisted.put(k, v);
+        });
+
+        // Replay all the entries and apply all the sequence ids updates
+        log.info("[{}] Replaying {} entries for deduplication", topic.getName(), managedCursor.getNumberOfEntries());
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        replayCursor(future);
+        return future;
+    }
+
+    /**
+     * Read all the entries published from the cursor position until the most recent and update the highest sequence id
+     * from each producer
+     *
+     * @param future
+     *            future to trigger when the replay is complete
+     */
+    private void replayCursor(CompletableFuture<Void> future) {
+        managedCursor.asyncReadEntries(100, new ReadEntriesCallback() {
+            @Override
+            public void readEntriesComplete(List<Entry> entries, Object ctx) {
+
+                for (Entry entry : entries) {
+                    ByteBuf messageMetadataAndPayload = entry.getDataBuffer();
+                    MessageMetadata md = Commands.parseMessageMetadata(messageMetadataAndPayload);
+
+                    String producerName = md.getProducerName();
+                    long sequenceId = md.getSequenceId();
+                    highestSequencedPushed.put(producerName, sequenceId);
+                    highestSequencedPersisted.put(producerName, sequenceId);
+
+                    md.recycle();
+                    messageMetadataAndPayload.release();
```

Comment:
either we release the `entry` or the buffer `messageMetadataAndPayload`, the effect is the same, though we cannot release both.

Reply:
> either we release the entry or the buffer messageMetadataAndPayload, the effect is the same

 Releasing [Entry](https://github.com/apache/incubator-pulsar/blob/master/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/EntryImpl.java#L149) also releases payload(`messageMetadataAndPayload`) but just releasing `payload` will not release `Entry` so, can you correct if it it's wrong but here we are not releasing `Entry` so, isn't that leak?

Reply:
I'll change it to call `entry.release()`. In any case it won't be a "leak", just a missed opportunity to recycling the entry instance.

Reply:
> In any case it won't be a "leak"

Yes, but if we don't recycle `EntryImpl` then don't we see leak with netty-detection report and it will keep creating new objects and will go out of memory.?

Reply:
No no, it's not that bad. When "leaking" objects from a recycler, they will just get normally GCed.